mypy_django_plugin/transformers/managers.py

```diff
@@ -183,112 +247,182 @@ def create_new_manager_class_from_from_queryset_method(ctx: DynamicClassDefConte
-    assert isinstance(derived_queryset_info, TypeInfo)
-
-    new_manager_info.line = ctx.call.line
-    new_manager_info.type_vars = base_manager_info.type_vars
-    new_manager_info.defn.type_vars = base_manager_info.defn.type_vars
-    new_manager_info.defn.line = ctx.call.line
-    new_manager_info.metaclass_type = new_manager_info.calculate_metaclass_type()
-    # Stash the queryset fullname which was passed to .from_queryset
-    # So that our 'resolve_manager_method' attribute hook can fetch the method from that QuerySet class
-    new_manager_info.metadata["django"] = {"from_queryset_manager": derived_queryset_fullname}
+    queryset_info = queryset_sym.node
+    assert isinstance(queryset_info, TypeInfo)
 
     if len(ctx.call.args) > 1:
         expr = ctx.call.args[1]
         assert isinstance(expr, StrExpr)
-        custom_manager_generated_name = expr.value
+        manager_class_name = expr.value
     else:
-        custom_manager_generated_name = base_manager_info.name + "From" + derived_queryset_info.name
+        manager_class_name = manager_base.name + "From" + queryset_info.name
 
-    custom_manager_generated_fullname = ".".join(["django.db.models.manager", custom_manager_generated_name])
-    base_manager_info.metadata.setdefault("from_queryset_managers", {})
-    base_manager_info.metadata["from_queryset_managers"][custom_manager_generated_fullname] = new_manager_info.fullname
+    manager_base_instance = fill_typevars(manager_base)
+    assert isinstance(manager_base_instance, Instance)
+    # Create a new `TypeInfo` instance for the manager type
+    new_manager_info = semanal_api.basic_new_typeinfo(
+        name=manager_class_name, basetype_or_fallback=manager_base_instance, line=ctx.call.line
+    )
+    new_manager_info.type_vars = manager_base.type_vars
+    new_manager_info.line = ctx.call.line
+    new_manager_info.defn.type_vars = manager_base.defn.type_vars
+    new_manager_info.defn.line = ctx.call.line
+    new_manager_info.metaclass_type = new_manager_info.calculate_metaclass_type()
 
-    # So that the plugin will reparameterize the manager when it is constructed inside of a Model definition
-    helpers.add_new_manager_base(semanal_api, new_manager_info.fullname)
+    merge_queryset_into_manager(
+        api=semanal_api,
+        manager_info=new_manager_info,
+        queryset_info=queryset_info,
+        manager_base=manager_base,
+        class_name=manager_class_name,
+    )
 
-    class_def_context = ClassDefContext(cls=new_manager_info.defn, reason=ctx.call, api=semanal_api)
-    self_type = fill_typevars(new_manager_info)
-    assert isinstance(self_type, Instance)
+    symbol_kind = semanal_api.current_symbol_kind()
+    # Annotate the module variable as `<Variable>: Type[<NewManager[Any]>]` as the model
+    # type won't be defined on variable level.
+    var = Var(
+        name=ctx.name,
+        type=TypeType(Instance(new_manager_info, [AnyType(TypeOfAny.from_omitted_generics)])),
```

Comment:
Why is this `Any`? Shouldn't we take the model from either the queryset or the manager (and potentially validate that they match)?

Reply:
I'm not sure, I'm kinda split on if we should display a model here. Since this is when we're outside of a model class definition, e.g.

```python
MyManager = Manager.from_queryset(MyQuerySet)
```

What I'm thinking about here is what the manager type is during runtime, model won't be set until the model metaclass is called. Which is basically inside model class definition. And, while I have never used it like that, same manager could be used on multiple models..

But perhaps you're right, maybe we should require model arguments for both queryset and manager to match. Since we're basically merging those two arguments into one.

Also, I think `AddManagers` will change the model argument if called on a different model? But perhaps that should also change?

Reply:
I was thinking it would make sense to inspect the generics of the manager and querysets (if possible). If they are different an error message should probably be emitted and in other case use the class from either?

And I'd think that if `AddManagers` sees a model argument that's not compatible with the one on the manager then an error should be emitted?

I'm okay with not adding this in this PR, but that's at least how I would expect this to behave. I see that it's really an edge case and probably not a big problem.

Reply:
Yeah, I've been looking in to exactly that. And it seems possible if we for `.from_queryset` traverse `.bases` (as those are `types.Instance`, we'll get access to `Instance.args`) and compare model arguments, then on mismatch emit an error.

This should probably happen for both `.from_queryset` and `AddManagers`, as you say.

Reply:
There's a case left to decide on, when model arguments aren't populated for `Queryset` or `Manager`. Either:

- We emit an error and enforce population of the model arg or,
- We then allow it to be any model, thus just populate it with whatever we got

I'm leaning towards the latter, since one can switch on enforcing via mypy config: `disallow_any_generics = true`.

For the first case, we could then establish some sort of "auto-fill". Depending on if 1 of the model arguments are populated, if both are missing `.from_queryset` results in `ManagerFromQueryset[Any]` while `AddManagers` just populates it with the model it's processing.

This is basically the table:

1. Manager and queryset model argument populated -> error if mismatch for both `.from_queryset` and `AddManagers`
2. Only manager model argument populated -> Populate queryset model argument with manager model argument
3. Only queryset model argument populated -> Populate manager model argument with manager model argument
4. No argument populated -> `.from_queryset` sets `Any` and `AddManagers` populates with the model it's processing.

I think this aligns with when using Django builtins e.g. `BaseManager.from_queryset(CustomQuerySet)`, since otherwise we'll enforce people to write custom managers to populate the argument, even though they only have/need a queryset defined. Like this:

```python
class SuperFluousManager(BaseManager[MyModel]):
    ...
```

What do you think?

Reply:
I've been looking in to this for a bit and think we should push for it to a future PR, as it's quite an isolated addition but generates quite a bit of lines. And the changes we have here are already large enough.

Reply:
I've pushed progress to a PR here: https://github.com/flaeppe/django-stubs/pull/29